virtcontainers/cgroups.go

```diff
@@ -0,0 +1,192 @@
+	}
+	return nil
+}
+
+func (s *Sandbox) destroyCgroups() error {
+	// first move all processes in subgroup to parent in case live process blocks
+	// deletion of cgroup
+	if s.cgroup == nil {
+		// no need to destroy
+		s.Logger().Warningf("cgroup not initialized, no need to destroy")
+		return nil
+	}
+
+	if err := s.cgroup.sandboxSub.MoveTo(s.cgroup.commonParent); err != nil {
+		return fmt.Errorf("failed to clear cgroup processes")
+	}
+
+	return s.cgroup.sandboxSub.Delete()
+}
+
+func (s *Sandbox) setupCgroups() error {
+	if s.cgroup == nil {
+		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+	}
+
+	resource, err := s.mergeSpecResource()
+	if err != nil {
+		return err
+	}
+
+	if err := s.applyCPUCgroup(resource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Sandbox) applyCPUCgroup(rc *specs.LinuxResources) error {
+	if s.cgroup == nil {
+		return fmt.Errorf("failed to setup uninitialized cgroup for sandbox")
+	}
+
+	// apply cpu constraint to vcpu cgroup
+	if err := s.cgroup.vcpuSub.Update(rc); err != nil {
+		return err
+	}
+
+	// when new container joins, new CPU could be hotplugged, so we
+	// have to query fresh vcpu info from hypervisor for every time.
+	tids, err := s.hypervisor.getThreadIDs()
+	if err != nil {
+		return err
+	}
+
+	if tids == nil {
+		// TODO: we only support qemu in this scenario, for unsupported hypervisor,
+		// just ignore and print a warning
+		s.Logger().Warning("failed to get thread ids from hypervisor")
+		return nil
+	}
+
+	// use Add() to add vcpu thread to s.cgroup, it will write thread id to
+	// `cgroup.procs` which will move all threads in qemu process to this cgroup
+	// immediately as default behaviour.
+	if len(tids.vcpus) > 0 {
+		if err := s.cgroup.sandboxSub.Add(cgroups.Process{
+			Pid: int(tids.vcpus[0]),
+		}); err != nil {
+			return err
+		}
+	}
+
+	for _, i := range tids.vcpus {
+		if i <= 0 {
+			continue
+		}
+
+		// In contrast, AddTask will write thread id to `tasks`
+		// After this, vcpu threads are in "vcpu" sub-cgroup, other threads in
+		// qemu will be left in parent cgroup untouched.
+		if err := s.cgroup.vcpuSub.AddTask(cgroups.Process{
+			Pid: int(i),
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Sandbox) mergeSpecResource() (*specs.LinuxResources, error) {
+	if s.config == nil {
+		return nil, fmt.Errorf("sandbox config is nil")
+	}
+
+	resource := &specs.LinuxResources{
+		CPU: &specs.LinuxCPU{},
+	}
+
+	for _, c := range s.config.Containers {
+		config, ok := c.Annotations[annotations.ConfigJSONKey]
+		if !ok {
+			s.Logger().WithField("container", c.ID).Warningf("failed to find config from container annotations")
+			continue
+		}
+
+		var spec specs.Spec
+		if err := json.Unmarshal([]byte(config), &spec); err != nil {
+			return nil, err
+		}
+
+		// TODO: how to handle empty/unlimited resource?
+		// maybe we should add a default CPU/Memory delta when no
+		// resource limit is given. -- @WeiZhang555
```

Comment:
Default cpu quota might conflict globally on the host if we exceed the total available CPU time. While users might set it to exceed, we should not make it happen unintentionally. docker/runc does not set default quota for containers either.

Reply:
The worry is, if there's two containers inside a POD, one with quota 6000, the other one with quota -1 (unlimited), total quota will be 6000.

That means with my calculation, one container with 0.6 core + one container with unlimited cores = 0.6 core.
Not sure if this can satisfy everyone, I can only suggest user, if you set a limit for one container, you should set limit for every container too...

Reply:
please file an issue